// enzyme-adapter-preact-pure/build/src/Adapter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var enzyme_1 = require("enzyme");
var preact_1 = require("preact");
var MountRenderer_1 = __importDefault(require("./MountRenderer"));
var ShallowRenderer_1 = __importDefault(require("./ShallowRenderer"));
var StringRenderer_1 = __importDefault(require("./StringRenderer"));
var compat_1 = require("./compat");
var util_1 = require("./util");
var preact10_rst_1 = require("./preact10-rst");
var preact8_rst_1 = require("./preact8-rst");
var Adapter = /** @class */ (function (_super) {
    __extends(Adapter, _super);
    function Adapter() {
        var _this = _super.call(this) || this;
        compat_1.addTypeAndPropsToVNode();
        _this.options = {
            // Prevent Enzyme's shallow renderer from manually invoking lifecycle
            // methods after a render. This manual invocation is needed for React
            // but not for the Preact adapter because we re-use the normal rendering
            // logic.
            lifecycles: {
                componentDidUpdate: {
                    onSetState: false,
                },
            },
        };
        // Work around a bug in Enzyme where `ShallowWrapper.getElements` calls
        // the `nodeToElement` method with undefined `this`.
        _this.nodeToElement = _this.nodeToElement.bind(_this);
        return _this;
    }
    Adapter.prototype.createRenderer = function (options) {
        switch (options.mode) {
            case 'mount':
                // The `attachTo` option is only supported for DOM rendering, for
                // consistency with React, even though the Preact adapter could easily
                // support it for shallow rendering.
                return new MountRenderer_1.default({ container: options.attachTo });
            case 'shallow':
                return new ShallowRenderer_1.default();
            case 'string':
                return new StringRenderer_1.default();
            default:
                throw new Error("\"" + options.mode + "\" rendering is not supported");
        }
    };
    Adapter.prototype.nodeToElement = function (node) {
        var _this = this;
        if (typeof node === 'string') {
            return node;
        }
        var childElements = node.rendered.map(function (n) { return _this.nodeToElement(n); });
        return preact_1.h.apply(void 0, __spreadArrays([node.type, node.props], childElements));
    };
    Adapter.prototype.nodeToHostNode = function (node) {
        if (typeof node === 'string') {
            // Returning `null` here causes `wrapper.text()` to return nothing for a
            // wrapper around a `Text` node. That's not intuitive perhaps, but it
            // matches the React adapters' behaviour.
            return null;
        }
        if (node.nodeType === 'host') {
            return node.instance;
        }
        else if (node.rendered.length > 0) {
            return this.nodeToHostNode(node.rendered[0]);
        }
        else {
            return null;
        }
    };
    Adapter.prototype.isValidElement = function (el) {
        if (el == null) {
            return false;
        }
        if (typeof el.type !== 'string' &&
            typeof el.type !== 'function' &&
            el.type !== null) {
            return false;
        }
        if (typeof el.props !== 'object' || el.props == null) {
            return false;
        }
        return true;
    };
    Adapter.prototype.createElement = function (type, props) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        return preact_1.h.apply(void 0, __spreadArrays([type, props], children));
    };
    Adapter.prototype.elementToNode = function (el) {
        var rstNodeFromElement = util_1.isPreact10()
            ? preact10_rst_1.rstNodeFromElement
            : preact8_rst_1.rstNodeFromElement;
        return rstNodeFromElement(el);
    };
    return Adapter;
}(enzyme_1.EnzymeAdapter));
exports.default = Adapter;
